### PR TITLE
Implement variant-based (C++17) message handling

### PR DESCRIPTION
**Pull request by netch80**

### CODE REVIEW

File: src/processor_cxx17/msg_handler.cxx [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+void MessageHandler::cbReceive(const std::vector<std::byte>& data)
+{
+    if (!data.size()) {
+        std::cerr << __func__ << ": buffer is empty" << std::endl;
+        return;
+    }
+
+    // this place should be optimized with PMR
+    container_t raw_msg(data);
+    std::unique_ptr<MessageVariant> msg_variant;
```
Comment (oleksiibidnichenko) [inline src/processor_cxx17/msg_handler.cxx:31]: this ugly unique_ptr, I really prefer to use std::optional for the case when you do not want to construct the object immediately

File: src/processor_cxx17/msg_handler.cxx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+void MessageHandler::HandleMessage(std::unique_ptr<MessageHandler::MessageVariant> msg)
+{
+    if (!msg) {
+        std::cerr << __func__ << ": No message to handle." << std::endl;
+        return;
+    }
+
+    // at this point message should be retrieved with std::variant
+    msg::MsgType msg_type;
+    std::visit([&](auto& arg) { msg_type = arg.GetMsgType(); }, *msg);
```
Comment (oleksiibidnichenko) [inline src/processor_cxx17/msg_handler.cxx:111]: I think here we already have arg and then later we use to get to obtain it once again. Of course, it is not critical, but it is a bit of an extra complicated approach.

File: include/processor_cxx17/msg_handler.hxx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    explicit MessageHandler(boost::asio::io_context &io_context);
+    ~MessageHandler() = default;
+
+    MessageHandler(const MessageHandler&) = delete;
+    MessageHandler& operator=(const MessageHandler&) = delete;
+
+    MessageHandler(MessageHandler&&) = delete;
+    MessageHandler& operator=(MessageHandler&&) = delete;
+
+    using container_t = std::vector<std::byte>;
```
Comment (oleksiibidnichenko) [inline include/processor_cxx17/msg_handler.hxx:29]: nice idea =)

File: include/processor_cxx17/msg_handler.hxx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    using MessageVariant = std::variant<
```
Comment (oleksiibidnichenko) [inline include/processor_cxx17/msg_handler.hxx:31]: IMHO this should be private/protected since this type is relevant only for the class hierarchy.

**Comment by netch80** [comment]
Cancelling due to basing on alternative version (#8).

**Pull request closed without merging**